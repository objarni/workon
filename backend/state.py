class State:
    def __init__(self):
        self.data = {}

    def add_client(self, alias, workon, timestamp):
        self.data[alias] = (workon, timestamp)
        to_delete = [alias for alias in self.data
                     if timestamp - self.data[alias][1] > 30]
        while len(to_delete) > 0:
            del self.data[to_delete[0]]
            del to_delete[0]

    def __repr__(self):
        result = ""
        for alias, data in self.data.items():
            workon, timestamp = data
            result += f"{alias} is working on {workon} since time={timestamp}\n"
        return result

    def get_working_clients(self):
        return [(alias, self.data[alias][0]) for alias in self.data]
