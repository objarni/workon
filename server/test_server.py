"""
- adding a new client in empty state
- adding a new client, when an old client is more than 30 s old
- adding a new client, when a prev client is less than 30 s old
"""
from approvaltests import verify, verify_as_json
from approvaltests.reporters import PythonNativeReporter

from state import State


def test_adding_client():
    state = State()
    state.add_client('Olof', 'rescue', timestamp=5)
    verify(repr(state), reporter=PythonNativeReporter())


def test_adding_second_client_within_30_seconds():
    state = State()
    state.add_client('Olof', 'rescue', timestamp=5)
    state.add_client('Tor', 'polarbear', timestamp=25)
    verify(repr(state), reporter=PythonNativeReporter())


def test_adding_second_client_after_31_seconds():
    state = State()
    state.add_client('Olof', 'rescue', timestamp=5)
    state.add_client('Tor', 'polarbear', timestamp=36)
    verify(repr(state), reporter=PythonNativeReporter())


def test_get_working_clients():
    state = State()
    state.add_client('Olof', 'rescue', timestamp=5)
    state.add_client('Tor', 'polarbear', timestamp=21)
    verify_as_json(state.get_working_clients(), reporter=PythonNativeReporter())
