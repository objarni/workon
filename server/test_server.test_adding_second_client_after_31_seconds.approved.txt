Tor is working on polarbear since time=36
